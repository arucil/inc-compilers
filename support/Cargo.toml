[package]
name = "support"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
assert_cmd = "2.0.4"
predicates = "2.1.1"
pretty_assertions = "1.2.1"
scopeguard = "1.1.0"
