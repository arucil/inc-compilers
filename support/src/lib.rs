pub mod compile_error;
pub mod range;
pub mod test_cli;

pub use self::compile_error::*;
pub use self::range::*;
pub use self::test_cli::*;
