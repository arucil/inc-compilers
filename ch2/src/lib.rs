#![allow(incomplete_features)]
#![feature(box_patterns, box_syntax, type_changing_struct_update)]

use support::CompileError;

pub mod pass;

pub fn compile(input: &str) -> Result<String, CompileError> {
  let prog = ast::parse(input)?;
  let prog = self::pass::partial_evaluation::partial_evaluate(prog);
  let prog = self::pass::uniquify::uniquify(prog);
  let prog = self::pass::remove_complex_operands::remove_complex_operands(prog);
  let prog = self::pass::explicate_control::explicate_control(prog);
  let prog = self::pass::instruction_selection::select_instruction(prog, false);
  let prog = self::pass::assign_home::assign_home(prog);
  let prog = self::pass::patch_instructions::patch_instructions(prog);
  let prog = self::pass::perilogue::add_perilogue(prog);

  Ok(prog.to_nasm(true))
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog = compile(
      r#"
(let
  ([x (read)]
   [y (+ 2 3)])
  (+ (- (read)) (+ y (- 2))))
    "#,
    )
    .unwrap();
    assert_snapshot!(prog);
  }
}
