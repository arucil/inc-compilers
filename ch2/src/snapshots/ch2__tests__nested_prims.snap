---
source: ch2/src/lib.rs
expression: prog
---
extern rt_print_int, rt_print_newline, rt_read_int
section .text

.start:
    call rt_read_int
    mov qword [rbp - 8], rax
    mov qword [rbp - 16], 5
    call rt_read_int
    mov qword [rbp - 24], rax
    mov rax, qword [rbp - 24]
    mov qword [rbp - 32], rax
    neg qword [rbp - 32]
    mov qword [rbp - 40], -2
    mov rax, qword [rbp - 16]
    add qword [rbp - 40], rax
    mov rax, qword [rbp - 32]
    add rax, qword [rbp - 40]
    jmp .conclusion

    global _start
_start:
    push rbp
    mov rbp, rsp
    sub rsp, 48
    jmp .start

.conclusion:
    mov rsp, rbp
    pop rbp
    mov rdi, rax
    call rt_print_int
    call rt_print_newline
    mov rax, 60
    mov rdi, 0
    syscall

